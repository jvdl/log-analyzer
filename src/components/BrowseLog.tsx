import { CHUNK_SIZE } from '../util/constants';

export function BrowseLog ({ logLines }: { logLines: string[] }) {
  if (!logLines?.length) {
    return null;
  }

  return (<>
    <h2>Browse the raw log</h2>
    <small>
      {
      logLines.length > CHUNK_SIZE
        ? <>Showing first {CHUNK_SIZE} lines</>
        : <>{logLines.length} lines</>
      }
    </small>
    <div className="raw-log">
      <pre>
        {logLines.slice(0,CHUNK_SIZE).map((line, index) => (
          <code key={index}>{line.trimStart()}{'\n'}</code>
        ))}
      </pre>
    </div>
  </>);
}
