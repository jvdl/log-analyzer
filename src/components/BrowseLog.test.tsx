import { BrowseLog } from './BrowseLog';
import renderer, { ReactTestRendererJSON, ReactTestRendererNode } from 'react-test-renderer';
import { describe, it, expect, vi } from 'vitest';

function isJsonNode(node?: ReactTestRendererNode): node is ReactTestRendererJSON {
  return node != null && typeof node !== 'string';
}

vi.mock('../util/constants', async (importOriginal) => {
  const mod = await importOriginal<typeof import('../util/constants')>();
  return {
    ...mod,
    // replace some exports
    CHUNK_SIZE: 5,
  };
});

describe(BrowseLog, () => {

  it('renders correctly with expected data', () => {
    const tree = renderer
      .create(<BrowseLog logLines={['line 1', 'line 2']} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
  });

  it('renders correctly with null data', () => {
    const tree = renderer
      // @ts-expect-error - passing null to a prop that expects an array
      .create(<BrowseLog logLines={null} />)
      .toJSON();
      expect(tree).toBeNull();
  });


  it('renders correctly when there are more lines than the chunk size', () => {
    const tree = renderer
      .create(<BrowseLog logLines={['line 1', 'line 2', 'line 3', 'line 4', 'line 5', 'line 6']} />)
      .toJSON() as ReactTestRendererJSON[] | null;

      // expect there to only be 5 lines in the 'raw-log'
      const rawLogNode = tree?.find(node => node.props.className === 'raw-log');
      const preNode = rawLogNode?.children?.[0];
      if (!isJsonNode(preNode)) {
        expect.fail('Expected preNode to be a ReactTestRendererJSON');
      }
      expect(preNode.children?.length).toBe(5);
  });

});
