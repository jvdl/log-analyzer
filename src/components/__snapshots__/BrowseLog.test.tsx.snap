// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BrowseLog > renders correctly with expected data 1`] = `
[
  <h2>
    Browse the raw log
  </h2>,
  <small>
    2
     lines
  </small>,
  <div
    className="raw-log"
  >
    <pre>
      <code>
        line 1
        

      </code>
      <code>
        line 2
        

      </code>
    </pre>
  </div>,
]
`;
